test/e2e/usersignup_test.go

```diff
@@ -0,0 +1,81 @@
+package e2e
+
+import (
+	"context"
+	"github.com/operator-framework/operator-sdk/pkg/test/e2eutil"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"testing"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/host-operator/pkg/apis"
+
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+)
+
+
+func TestUserSignup(t *testing.T) {
+	userSignupList := &v1alpha1.UserSignupList{}
+
+	err := framework.AddToFrameworkScheme(apis.AddToScheme, userSignupList)
+	require.NoError(t, err, "failed to add custom resource scheme to framework")
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup()
+
+	err = ctx.InitializeClusterResources(&framework.CleanupOptions{TestContext: ctx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	require.NoError(t, err, "failed to initialize cluster resources")
+	t.Log("Initialized cluster resources")
+
+	namespace, err := ctx.GetNamespace()
+	require.NoError(t, err, "failed to get namespace where operator needs to run")
+
+	f := framework.Global
+	//client := f.Client.Client
+
+	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "host-operator", 1, operatorRetryInterval, operatorTimeout)
+	require.NoError(t, err, "failed while waiting for host operator deployment")
+
+	t.Log("host-operator is ready and in a running state")
+
+	// Create user signup
+	t.Logf("Creating UserSignup with namespace %s", namespace)
+	userSignup := newUserSignup(namespace, "foo")
+	err = f.Client.Create(context.TODO(), userSignup, &framework.CleanupOptions{TestContext: ctx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	require.NoError(t, err)
+	t.Logf("user signup '%s' created", userSignup.Name)
+
+	// Confirm the UserSignup was created
+	err = verifyResources(t, f, namespace, userSignup)
+	require.NoError(t, err)
+
+	// Delete the User Signup
```

Comment:
We need more tests here.
For three states of configuration:
- manual approval
- auto-approval
- no configuration exist (so, the default manual approval should be used)

We need to check all the cases for the configurations from above:
- Create a signup with approved not set
- Create a signup with approved == false
- Create a signup with approved == true
- Update an existing signup (from approved == false to approved == true)

Also tests for:
- Target cluster specified in userSignup
- Target cluster is not set and picked by the controller
- Delete the existing UserSignup (garbage collector should automatically delete the MUR)
- Check the expected status of UserSignup (the status will depend on the use case)

Reply:
Most of these are done now, just the last few need to be taken care of.  I have one challenge in testing an existing signup where I set approved from `false` to `true` - the `waitForUserSignup()` function currently only waits for the resource to be created in the first place.  I need to somehow wait for the controller to finish reconciling and only then read the updated `UserSignup` instance.  Any ideas on how I might easily do this?

Reply:
You can't wait for the reconciling to finish in e2e tests but you don't have to. This is what you can do instead:
- Create a UserSignup. Wait for it to get created with expected status (should be pending approval). It's important to wait for the condition(s) set to the expected values because UserSignup is first created and then we update the status during reconcile. Check that no MUR created.
- Approve it. Wait for the status condition Complete==true. Check that the corresponding MUR created.

Reply:
All done now